### PR TITLE
stm32: add supports for configure MCO when startup

### PR DESCRIPTION
**Pull request by qianfan-Zhao**
tested in nucleo_f070rb (MCO  pin is PA8)

```
CONFIG_CLOCK_STM32_MCO1_SRC_LSI=1
CONFIG_CLOCK_STM32_MCO1_DIV=1
```

### CODE REVIEW

**Comment by zhaozhongchen** [comment]
@erwango @ydamigos please have a look.

I had only tested this pr on nucleo_f070rb, and haven't add pinmux for stm32f1, f3, f4, f7, l0, l4. Will do that later.

**Comment by erwango** [comment]
@QianFan-Zhao , thanks for this change.
As this change is impacting all series and boards, it will take me sometime to review and test in deep.
I'll try to get back to you next week on it.

**Comment by zhaozhongchen** [comment]
Add 'STM32_OSPEEDR_VERY_HIGH_SPEED' flag to MCO pin and force push

**Comment by zhaozhongchen** [comment]
ping @erwango

File: drivers/clock_control/stm32_ll_clock.c [hunk 1]
```diff
@@ -25,6 +25,54 @@
...
 #endif /* CONFIG_SOC_SERIES_STM32F0X  */
 
+#define _mco1_prescaler(v) LL_RCC_MCO1_DIV_ ## v
+#define mco1_prescaler(v) _mco1_prescaler(v)
+
+#define _mco2_prescaler(v) LL_RCC_MCO2_DIV_ ## v
+#define mco2_prescaler(v) _mco2_prescaler(v)
+
+#if CONFIG_CLOCK_STM32_MCO1_SRC_NOCLOCK
+	#define LL_RCC_MCO1SOURCE		LL_RCC_MCO1SOURCE_NOCLOCK
```
Comment (erwango) [inline drivers/clock_control/stm32_ll_clock.c]: "LL_RCC_MCO1SOURCE": You could not add your own defines using STM32Cube namespace (LL_RCC), unless it is a redefinition of existing define (for a good reason).
We should keep clear separation between STM32Cube definitions and the one created for Zephyr driver.
"MCO1_SOURCE" should be sufficient.

File: drivers/clock_control/stm32_ll_clock.c [hunk 2]
```diff
@@ -375,6 +423,24 @@ static int stm32_clock_control_init(struct device *dev)
 
 #endif /* CONFIG_CLOCK_STM32_SYSCLK_SRC_... */
 
+#ifndef CONFIG_CLOCK_STM32_MCO1_SRC_NOCLOCK
```
Comment (erwango) [inline drivers/clock_control/stm32_ll_clock.c]: Please put this under a specific function (stm32clock_control_mco_init for instance)
stm32_clock_control_init is already rather long, we should not add too much things.

Comment (erwango) [reply drivers/clock_control/stm32_ll_clock.c]: Besides, according to Reference manual, it should be done before PLL activation.

File: drivers/clock_control/stm32_ll_clock.c [hunk 3]
```diff
@@ -375,6 +423,24 @@ static int stm32_clock_control_init(struct device *dev)
...
+#ifndef CONFIG_CLOCK_STM32_MCO1_DIV
+	LL_RCC_ConfigMCO(LL_RCC_MCO1SOURCE);
```
Comment (erwango) [inline drivers/clock_control/stm32_ll_clock.c]: Please add a comment that MCO does not activate requested source and on matching source activation should be checked

**Review by erwango — CHANGES_REQUESTED** [review]
Looks nice.
Some initial comments.

I wonder if it is reasonnable to implement for all series right away, as this will request lot of test effort.

**Comment by zhaozhongchen** [comment]
@erwango stm32 all series MCO is quiet similer, so I want add them at all.

**Comment by erwango** [comment]
> @erwango stm32 all series MCO is quiet similer, so I want add them at all.

@zhaozhongchen , I agree this makes sense to take advantage of your developments, but we need to test on all series

**Comment by zhaozhongchen** [comment]
> but we need to test on all series

@erwango Should I remove all soc except STM32F0 to make the test easier?

File: drivers/clock_control/Kconfig.stm32 [hunk 4]
```diff
@@ -99,6 +99,94 @@ config CLOCK_STM32_PLL_SRC_PLL2
 
 endchoice
 
+# Will add others SOC supports later
```
Comment (erwango) [inline drivers/clock_control/Kconfig.stm32]: Please remove this line

File: drivers/clock_control/stm32_ll_clock.c [hunk 5]
```diff
@@ -198,6 +226,31 @@ static void stm32_clock_switch_to_hsi(u32_t ahb_prescaler)
 	}
 }
 
+/*
+ * MCO configure doesn't active requested clock source,
```
Comment (erwango) [inline drivers/clock_control/stm32_ll_clock.c]: Then we should activate source clock in stm32_clock_control_mco_init,
at least a check clock is enabled would be required.

File: drivers/clock_control/stm32_ll_clock.c [hunk 6]
```diff
@@ -210,6 +263,9 @@ static int stm32_clock_control_init(struct device *dev)
 	/* Some clocks would be activated by default */
 	config_enable_default_clocks();
 
+	/* configure MCO1/MCO2 based on Kconfig */
+	stm32_clock_control_mco_init();
```
Comment (erwango) [inline drivers/clock_control/stm32_ll_clock.c]: Please move this call to the end of the function

**Review by erwango — CHANGES_REQUESTED** [review]
@qianfan-Zhao , sorry for delay in reviewing your change.
I have some few additional comments, then we're good for merge

**Comment by galak** [comment]
@qianfan-Zhao Any plans to update?

**Comment by qianfan-Zhao** [comment]
@galak  Yes I will update this pr in recent days

**Comment by qianfan-Zhao** [comment]
@galak  Yes I will update this pr in recent days

**Review by erwango — CHANGES_REQUESTED** [review]
Some last comments and we're good.

**Comment by erwango** [comment]
^^@qianfan

File: drivers/clock_control/Kconfig.stm32 [hunk 7]
```diff
@@ -116,6 +116,92 @@ config CLOCK_STM32_PLL_SRC_PLL2
...
+	help
+	  MCO2 output disabled, no clock on MCO2
+
+config CLOCK_STM32_MCO2_SRC_SYSCLK
+	bool "SYSCLK"
+	depends on SOC_SERIES_STM32F4X
+	help
+	  Use SYSCLK as source of MCO2
+
+config CLOCK_STM32_MCO2_SRC_PLLI2S
```
[the remark this answers is not in the document]
Comment (qianfan-Zhao) [reply drivers/clock_control/Kconfig.stm32]: On RM0090 Rev17 Section 6.2.10:
```
You can output four different clock sources onto the MCO2 pin (PC9) using the
configurable prescaler (from 1 to 5):
– HSE clock
– PLL clock
– System clock (SYSCLK)
– PLLI2S clock
```

Which version of RM0090 you used?

**Comment by qianfan-Zhao** [comment]
@erwango Sorry for my later update.

File: drivers/clock_control/Kconfig.stm32 (hunk 7 above)
[the remark this answers is not in the document]
Comment (erwango) [reply drivers/clock_control/Kconfig.stm32]: I was referring to the clock tree.
If PLLI2S is used elsewhere, I'm good with that too.

**Review by erwango — APPROVED** [review]

**Comment by erwango** [comment]
@galak ^^

**Pull request merged**